File: Trading bot.java
const { EClient } = require('ib-api')
const { EWrapper } = require('ib-api')
const { Contract } = require('ib-api')
const { Order } = require('ib-api')
const { BarData } = require('ib-api')
const client = new EClient(new MyWrapper())

class MyWrapper extends EWrapper {
  constructor() {
    super()
    this.nextOrderId = null
    this.orderStatus = {}
    this.historicalData = []
  }

  nextValidId(orderId) {
    super.nextValidId(orderId)
    this.nextOrderId = orderId
  }

  orderStatus(orderId, status, filled, remaining, avgFillPrice, permId, parentId, lastFillPrice, clientId, whyHeld, mktCapPrice) {
    this.orderStatus[orderId] = status
  }

  openOrder(orderId, contract, order, orderState) {
    console.log("Open Order:", orderId, contract.symbol, contract.secType, order.action, order.orderType, order.totalQuantity, orderState.status)
  }

  historicalData(reqId, bar) {
    this.historicalData.push(bar)
  }
}

function createContract(symbol, secType, exchange, currency, expiry="", strike=0, right="") {
  const contract = new Contract()
  contract.symbol = symbol
  contract.secType = secType
  contract.exchange = exchange
  contract.currency = currency
  contract.lastTradeDateOrContractMonth = expiry
  contract.strike = strike
  contract.right = right
  return contract
}

function createOrder(action, quantity, orderType, price) {
  const order = new Order()
  order.action = action
  order.orderType = orderType
  order.totalQuantity = quantity
  order.lmtPrice = price
  return order
}

function placeOrder(orderId, contract, order) {
  client.placeOrder(orderId, contract, order)
}

function cancelOrder(orderId) {
  client.cancelOrder(orderId)
}

function reqHistoricalData(reqId, contract, endDateTime, durationStr, barSizeSetting, whatToShow, useRTH, formatDate=2, keepUpToDate=false, chartOptions=null) {
  client.reqHistoricalData(reqId, contract, endDateTime, durationStr, barSizeSetting, whatToShow, useRTH, formatDate, keepUpToDate, chartOptions)
}

async function main() {
  await client.connect("127.0.0.1", 7497, 0)

  // Create the contract for the option being traded
  const contract = createContract("QQQ", "OPT", "SMART", "USD", "20230303", 400, "CALL")

  // Request historical data for the option
  reqHistoricalData(1, contract, "", "1 D", "5 mins", "TRADES", 0)
}

main()

File: trading bot.py
from ibapi.client import EClient
from ibapi.wrapper import EWrapper
from ibapi.contract import Contract
from ibapi.order import Order
from ibapi.common import BarData
import time

class MyWrapper(EWrapper):
    def __init__(self):
        self.nextOrderId = None
        self.orderStatus = {}
        self.historicalData = []

    def nextValidId(self, orderId:int):
        super().nextValidId(orderId)
        self.nextOrderId = orderId

    def orderStatus(self, orderId:int, status:str, filled:float, remaining:float, avgFillPrice:float, permId:int, parentId:int, lastFillPrice:float, clientId:int, whyHeld:str, mktCapPrice:float):
        self.orderStatus[orderId] = status

    def openOrder(self, orderId:int, contract:Contract, order:Order, orderState):
        print("Open Order:", orderId, contract.symbol, contract.secType, order.action, order.orderType, order.totalQuantity, orderState.status)

    def historicalData(self, reqId:int, bar:BarData):
        self.historicalData.append(bar)

class MyClient(EClient):
    def __init__(self, wrapper):
        EClient.__init__(self, wrapper)

    def createContract(self, symbol, secType, exchange, currency, expiry="", strike=0, right=""):
        contract = Contract()
        contract.symbol = symbol
        contract.secType = secType
        contract.exchange = exchange
        contract.currency = currency
        contract.lastTradeDateOrContractMonth = expiry
        contract.strike = strike
        contract.right = right
        return contract

    def createOrder(self, action, quantity, orderType, price):
        order = Order()
        order.action = action
        order.orderType = orderType
        order.totalQuantity = quantity
        order.lmtPrice = price
        return order

    def placeOrder(self, orderId, contract, order):
        self.placeOrder(orderId, contract, order)

    def cancelOrder(self, orderId):
        self.cancelOrder(orderId)

    def reqHistoricalData(self, reqId, contract, endDateTime, durationStr, barSizeSetting, whatToShow, useRTH, formatDate=2, keepUpToDate=False, chartOptions=None):
        self.reqHistoricalData(reqId, contract, endDateTime, durationStr, barSizeSetting, whatToShow, useRTH, formatDate, keepUpToDate, chartOptions)

def main():
    wrapper = MyWrapper()
    client = MyClient(wrapper)
    client.connect("127.0.0.1", 7497, 0)

    # Create the contract for the option being traded
    contract = client.createContract("TSLA", "OPT", "SMART", "USD", "20230303", 800, "CALL")

    # Request historical data for the option
    client.reqHistoricalData(1, contract, "", "1 D", "5 mins", "TRADES", 0)

if __name__ == "__main__":
    main()
